### PR TITLE
[FEATURE] Add Support for Returning Parameters and Metrics as DataAssistantResult class

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* When the underlying Rule-Based Profiler (RBP) is executed for a `DataAssistant` object, it returns the immutable `DataAssistantResult` dataclass, which contains the essential `DataAssistant` accessors (metrics, expectations, etc.).
* The `VolumeDataAssistant` has been updated to utilize the new `DataAssistantResult` object.
* Updated functionality: the `profiler_config` component of `ExpectationSuite` metadata citation now contains the effective Profiler configuration, which may not be the same as the one corresponding to the RBP `YAML` configuration, accessed by `ProfilerStore` from persistent storage; the effective Profiler configuration incorporates overrides, reconciled at runtime.
* All relevant unit tests have been updated.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-735/GREAT-597
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py [hunk 1]
```diff
@@ -125,6 +125,24 @@ def test_alice_profiler_user_workflow_single_batch(
...
         include_citation=True,
     )
+
+    assert (
+        expectation_suite.expectations
+        == alice_columnar_table_single_batch["expected_expectation_suite"].expectations
+    )
+
+    """
+    Deleting some "parameter_builders" from both actual and expected configurations, because nested dictionaries are
```
Comment (anthonyburdi) [inline tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py:135]: I'm not a huge fan of this, but understand that it is some acceptable tech debt given that to change it requires a larger fix to our serialization mechanisms.

File: great_expectations/rule_based_profiler/data_assistant/data_assistant.py [hunk 2]
```diff
@@ -160,7 +165,16 @@ def run(self) -> None:
...
+        return DataAssistantResult(
+            profiler_config=self.profiler.config,
+            metrics=self.get_metrics(),
+            expectation_configurations=self.get_expectation_configurations(),
+            expectation_suite=self.get_expectation_suite(
+                expectation_suite=expectation_suite,
+                expectation_suite_name=expectation_suite_name,
+                include_citation=include_citation,
+            ),
+        )
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/data_assistant/data_assistant.py:177]: ❤️

**Review by anthonyburdi — APPROVED** [review]
LGTM!

File: tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py (hunk 1 above)
Comment (alexsherstinsky) [reply tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py:135]: @anthonyburdi I kept trying for hours and could not make it work, even though they were equal as confirmed manually.  At least there is a full test in the VolumeDataAssistant test module.  Thank you.

File: great_expectations/rule_based_profiler/types/data_assistant_result.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+from dataclasses import asdict, dataclass
+from typing import Any, Dict, List
+
+from great_expectations.core import ExpectationConfiguration, ExpectationSuite
+from great_expectations.core.util import convert_to_json_serializable
+from great_expectations.rule_based_profiler.types import Domain
+from great_expectations.types import SerializableDictDot
+
+
+@dataclass(frozen=True)
```
Comment (Shinnnyshinshin) [inline great_expectations/rule_based_profiler/types/data_assistant_result.py:10]: ❤️

File: tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py [hunk 4]
```diff
@@ -461,6 +479,23 @@ def test_bobby_profiler_user_workflow_multi_batch_row_count_range_rule_and_colum
...
+    extremely difficult to match, due to complex structure, containing dictionaries, lists of dictionaries, lists of
+    primitive types, and other iterables, presented in random sort order at various levels of configuration hierarchy.
+    """
+    profiled_expectation_suite.meta["citations"][0]["profiler_config"]["rules"][
+        "my_rule_for_timestamps"
+    ].pop("parameter_builders")
+    fixture_expectation_suite.meta["citations"][0]["profiler_config"]["rules"][
+        "my_rule_for_timestamps"
+    ].pop("parameter_builders")
```
Comment (Shinnnyshinshin) [inline tests/integration/profiling/rule_based_profilers/test_profiler_user_workflows.py:497]: 😅

**Review by Shinnnyshinshin — COMMENTED** [review]
Thank you very much for this @alexsherstinsky. Excited to see it move forward

**Pull request merged**